Update Jupyter notebooks and associated codebase [RMG-Py]

jonwzheng opened the pull request:
### Motivation or Problem
Our Jupyter notebooks have fallen out-of-date.
In the process of fixing the notebooks, this has revealed some problems with the codebase as well.

Some parts of this PR are dependent on others:
* Uncertainty estimation depends on updates to the kinetic uncertainty estimate methods (#2787)
* Also depend on updates to Python (as `muq` is currently blocked.) #2796 and possibly others

### Description of Changes
Updated the following:
- [x] `cantera_sensitivity_comparison.ipynb` fixed
- [x] `cantera_simulation.ipynb` fixed codebase (amazingly, there was a glitch from recent changes to the surface code making this completely not work at all! The code is now refactored.) 
- [x] `combustion_model_and_ignition_delay_demo` fixed
- [x] `estimate_solvation_thermo_and_search_available_solvents`  OK
- [x] `find_parameter_sources_and_assign_uncertainties`  Fixed. Many thanks for working on #2787. 
- [x] `fragment_reattachment_example` Big changes to this + `fragment_utils`. Basically moved most of the class defs here to the `utils`, because the `utils` code was weirdly coupled to the `FragList` class which was only ever defined in the Jupyter notebook. Fixed, so it can now run (somewhat) reliably, and each file independently makes sense. This code is a little weird, as the reattachment is recursive AND the grinding is random, so you can sometimes run into recursion errors.
- [x] `generate_reactions` OK
~~- [x] `global_uncertainty` Doesn't work without `muq` which can't be installed.~~ see: https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2815#issuecomment-3025040934
- [x] `kinetics_library_to_training` OK
- [x] `local_uncertainty`  Fixed in large part to #2787. 
- [x] `mechanism_analyzer` fixed
- [x] `regression_test` fixed (notebook + codebase bugs)
- [x] `temperature_dependent_solvation_free_energy` fixed (and updated docstrings) 

RMS-related examples in `/examples`:
- `{sensitivity_analysis, simulation_flux_rop, tolerance}_superminimal_RMS.ipynb` I haven't touched; needs to be updated for new way of invoking Julia 
- `simulate_ch4o2cat_RMS.ipynb` see above

These are addressed in separate PR #2829.

### Other
We may want to add these to our testing suite

jonwzheng committed:
update temperature-dependent solvation notebook

jonwzheng committed:
initial changes to regression test example

jonwzheng committed:
update kinetic mechanism analyzer

jonwzheng committed:
fix typo in observablesregression fix

jonwzheng committed:
update cantera sensitivity comparison

jonwzheng committed:
fix keywords

jonwzheng committed:
update cantera notebooks

jonwzheng committed:
update np.float from cksv parser

jonwzheng committed:
update combustion demo

jonwzheng committed:
initial fix to parameter uncertainty assignment

jonwzheng committed:
clear local uncertainty

jonwzheng committed:
update regression test

jonwzheng commented:
I ran into trouble with getting the uncertainty stuff to work. @mjohnson541 would you please be able to take a look? I made a change to `family.py` but feel free to discard if that's the wrong direction...

jonwzheng commented:
Actually, scratch that. I believe the uncertainty notebooks will be fixed/addressed in #2787  so no need for changes on in this PR

sevyharris commented:
> > Actually, scratch that. I believe the uncertainty notebooks will be fixed/addressed in #2787 so no need for changes on in this PR
> 
> #2787 is now merged.

The local uncertainty notebook/underlying code may still need a few changes to run again, but I will check and take care of it. At the very least, the changes from #2787 will make this a lot easier.

jonwzheng committed:
fix circular import error when running generateFluxDiagram

jonwzheng committed:
fix local uncertainty notebook

Also add the two small simulation outputs so that the user doesn't have
to rerun the job.

jonwzheng committed:
fix find_parameter_sources notebook

jonwzheng committed:
Fix fragment notebook & utils

The previous implementation of the fragment notebook/utils don't work
correctly. The utils didn't work independently outside of the notebook.
This commit moves the FragmentList definition to the utils code, and
just imports everything from the notebook. It also fixes some other
errors in the codebase.

jonwzheng committed:
keep full fragment reattachment example output

jonwzheng commented:
The Julia notebooks were updated thanks to Matt in #2829, so this PR is ready to go (absent `global_uncertainty` which will probably need an update to Python 3.11, which might take a bit more time).

So, this PR is ready for review

jonwzheng commented:
Looking to get this reviewed and merged in before it's forgotten. @sevyharris if you don't have capacity, I've tagged @KnathanM and @BonhyeokKoo to take a look

JacksonBurns commented:
The issue with `muq` will be resolved with the merger of #2796

jonwzheng committed:
Fix previous bug with thermo data plotting

Previous implementation multiplied list by 418400 length rather than
scaling each item by said amount. This caused the plot to show basically
the same value plotted a bunch of times because it was stretched out.

jonwzheng commented on rmgpy/tools/canteramodel.py:
Thanks for catching this Sevy. latest commit addresses this (in a slightly different way) and I tested the notebook.

<img width="747" height="433" alt="image" src="https://github.com/user-attachments/assets/4185ecce-2fb1-4828-806b-cd30408b16a6" />

jonwzheng committed:
Fix typos in notebooks

Addit'l errors were either uncaught or perhaps introduced in previous
commits. Thanks @sevyharris for catching these!

jonwzheng committed:
Update ckcsvparser float types

Previous commit updated np.float to np.float64
This now also updates the generic python 'float' to np.float64 in that
same file

sevyharris commented:
I got the fragment example to run for a tiny model, but it doesn't show a nice distribution of all the large molecules. 

I'm guessing the saved model just needed A LOT more time and compute power to solve.

I propose we replace the saved model with a smaller version, and if that still takes too long to simulate, we save the results in a csv or npy then leave a cell that says "this is how you'd run the simulation if you had a lot of time and CPUs but here's an answer you can load in seconds."

@jonwzheng is it okay for me to push to this branch to make the above changes?

jonwzheng commented:
That sounds fine to me!

sevyharris committed:
Use smaller model for fragment reattachment example

The simulation for the fragment reattachment example is taking way too
long (more than a minute per solver step), so this replaces the example
mechanism (~400 species ~7000 reactions) with a much smaller example (48
species 257 reactions) that still shows a reasonable distribution of
large molecules

sevyharris reviewed (approved):
Someone should double check the changes I just made to the fragment example, but otherwise I think this is good to go.

JacksonBurns reviewed (approved):
@sevyharris thanks for making those changes - this looks much easier to run in a reasonable amount of time, allowing users to focus on actually reading what's going on. Going to update this branch and set to merge.
